src/server.ts

```diff
@@ -114,4 +116,63 @@ export default class SocketServer extends EventEmitter {
 
     return this.deno_server;
   }
+
+  /**
+   * @description
+   *    Handles websocket secure connection. After a successful connection,
+   *    the client will be added to EventEmitter.clients and the server will
+   *    start listening to events.
+   *
+   * @param HTTPOptions options
+   *
+   * @return Promise<DenoServer>
+   */
+  public async runTLS(options: HTTPSOptions): Promise<DenoServer> {
+    if (options.hostname) {
+      this.hostname = options.hostname;
+    }
+
+    if (options.port) {
+      this.port = options.port;
+    }
+
+    this.deno_server = serveTLS(options);
+
+    (async () => {
+      for await (const req of this.deno_server) {
+        const { conn, r: bufReader, w: bufWriter, headers } = req;
+
+        acceptWebSocket({
+          conn,
+          headers,
+          bufReader,
+          bufWriter,
+        })
+          .then(async (socket: WebSocket): Promise<void> => {
+            const clientId = conn.rid;
+            super.addClient(clientId, socket);
+
+            try {
+              for await (const ev of socket) {
+                if (ev instanceof Uint8Array) {
+                  await super.checkEvent(ev, clientId);
+                } else if (isWebSocketCloseEvent(ev)) {
+                  await super.removeClient(clientId);
+                }
+              }
+            } catch (e) {
+              if (!socket.isClosed) {
+                await socket.close(1000).catch(console.error);
+                await super.removeClient(clientId);
+              }
+            }
+          })
+          .catch((err: Error): void => {
+            console.error(`failed to accept websocket: ${err}`);
+          });
+      }
+    })();
+
+    return this.deno_server;
+  }
```

Comment:
can we separate this func out and runTLS and run can just invoke one func?

Reply:
Shouldn't be hard. Pseudocode:
```typescript
if (this.config.certFile != null && this.config.keyFile != null) {
  this.deno_server = serveTLS(options);
} else {
  this.deno_server = serve(options);
}
```

Reply:
we would need to do the same in drash

Reply:
Let’s make an issue about this to investigate more because I feel like we should keep run and runTLS to be explicit